numpy/random/mt19937.pyx

```diff
@@ -128,121 +131,74 @@ cdef class MT19937:
 
     """
     cdef mt19937_state rng_state
-    cdef bitgen_t _bitgen
-    cdef public object capsule
-    cdef object _ctypes
-    cdef object _cffi
-    cdef public object lock
 
     def __init__(self, seed=None):
-        self.seed(seed)
-        self.lock = Lock()
-
+        BitGenerator.__init__(self, seed)
+        if isinstance(seed, SeedSequence):
+            val = self._seed_seq.generate_state(RK_STATE_LEN, np.uint32)
+            # MSB is 1; assuring non-zero initial array
+            self.rng_state.key[0] = 0x80000000UL 
+            for i in range(1, RK_STATE_LEN):
+                self.rng_state.key[i] = val[i]
+            self.rng_state.pos = i
+        else:
+            self.legacy_seeding(seed)
         self._bitgen.state = &self.rng_state
         self._bitgen.next_uint64 = &mt19937_uint64
         self._bitgen.next_uint32 = &mt19937_uint32
         self._bitgen.next_double = &mt19937_double
         self._bitgen.next_raw = &mt19937_raw
 
-        self._ctypes = None
-        self._cffi = None
-
-        cdef const char *name = "BitGenerator"
-        self.capsule = PyCapsule_New(<void *>&self._bitgen, name, NULL)
-
-    # Pickling support:
-    def __getstate__(self):
-        return self.state
-
-    def __setstate__(self, state):
-        self.state = state
-
-    def __reduce__(self):
-        from ._pickle import __bit_generator_ctor
-        return __bit_generator_ctor, (self.state['bit_generator'],), self.state
-
-    def random_raw(self, size=None, output=True):
-        """
-        random_raw(self, size=None)
-
-        Return randoms as generated by the underlying BitGenerator
-
-        Parameters
-        ----------
-        size : int or tuple of ints, optional
-            Output shape.  If the given shape is, e.g., ``(m, n, k)``, then
-            ``m * n * k`` samples are drawn.  Default is None, in which case a
-            single value is returned.
-        output : bool, optional
-            Output values.  Used for performance testing since the generated
-            values are not returned.
-
-        Returns
-        -------
-        out : uint or ndarray
-            Drawn samples.
-
-        Notes
-        -----
-        This method directly exposes the the raw underlying pseudo-random
-        number generator. All values are returned as unsigned 64-bit
-        values irrespective of the number of bits produced by the PRNG.
-
-        See the class docstring for the number of bits returned.
+    def legacy_seeding(self, seed=None):
```

Comment:
This function could also be cdef and then called in RandomState so that it wouldn't appear to Python users. Would make the APIs superficially identical.

Reply:
I think that's right. We do need an *instancemethod* that reseeds an existing `MT19937` instance in order to support `RandomState.seed()`. Making it a `cdef` instancemethod would be good to hide it from the Python-level interface. Calling it `legacy_reseed()` might clarify what it's doing better.

But we also need a way to _create_ `MT19937` instances with the legacy seeding (providing that it is not the _default_ way). At minimum, we need to do it inside of `RandomState.__init__()`. It would be _permissible_ to make that an implementation detail of `RandomState`. However, I thought that it might be a good idea to expose that at the Python level with a `@classmethod` so that people could get more comfortable with the rest of the changes if they could exactly reproduce the same MT state with `Generator`. I could be overthinking that, though. Maybe we could add that `@classmethod` in 1.18 if people really ask for it. In the meantime, they could just grab it off of `np.random.RandomState(12345)._bit_generator`.

Reply:
One method to achieve the second goal would be to instantize any MT19937 (e.g., `MT19937()` which would use a `SeedSequence`) and then to call `legacy_reseed` in `__init__`. 

In this model, `legacy_reseed` would exist exclusively for `RandomState`.  If cdef is too much of a pain, then it probably should be `_legacy_reseed` to prevent trivial autocomplete, and to inidcate that it is an implementation detail. Docstring could say as much too.

Reply:
There are tests that check `Generator(s)` is equivalent to `RandomState(MT19937(s))`. Also tests that call `Generator(MT19337(s))` now go through a `SeedSequence` so the result changes.

Reply:
So change that to `Generator(RandomState(s)._bit_generator)`

Reply:
Or go ahead and add the `@classmethod` and use `Generator(MT19937.legacy_seeding(s))` in those tests. That could be enough justification for adding the public `@classmethod`, IMO.

Reply:
If I already need to change the tests, I will change the desired result.